// @nrwl/schematics/src/command-line/format.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var child_process_1 = require("child_process");
var path = require("path");
var resolve = require("resolve");
var shared_1 = require("./shared");
function format(args) {
    var command = args[0];
    var patterns;
    var rest;
    try {
        if (args.length === 1) {
            patterns = ['"{apps,libs}/**/*.ts"'];
            rest = [];
        }
        else {
            var p = shared_1.parseFiles(args.slice(1));
            patterns = p.files.filter(function (f) { return path.extname(f) === '.ts'; });
            rest = p.rest;
            var libsAndApp = rest.filter(function (a) { return a.startsWith('--libs-and-apps'); })[0];
            if (libsAndApp) {
                patterns = getPatternsFromApps(patterns);
            }
        }
    }
    catch (e) {
        printError(command, e);
        process.exit(1);
    }
    switch (command) {
        case 'write':
            write(patterns);
            break;
        case 'check':
            check(patterns);
            break;
    }
}
exports.format = format;
function getPatternsFromApps(affectedFiles) {
    var roots = shared_1.getAppRoots(shared_1.getAffectedApps(affectedFiles));
    if (roots.length === 0) {
        return [];
    }
    else if (roots.length === 1) {
        return ["\"" + roots[0] + "/**/*.ts\""];
    }
    else {
        return ["\"{" + roots.join(',') + "}/**/*.ts\""];
    }
}
function printError(command, e) {
    console.error("Pass the SHA range, as follows: npm run format:" + command + " -- SHA1 SHA2.");
    console.error("Or pass the list of files, as follows: npm run format:" + command + " --files=\"libs/mylib/index.ts,libs/mylib2/index.ts\".");
    console.error(e.message);
}
function write(patterns) {
    if (patterns.length > 0) {
        child_process_1.execSync("node " + prettierPath() + " --single-quote --print-width 120 --write " + patterns.join(' '), {
            stdio: [0, 1, 2]
        });
    }
}
function check(patterns) {
    if (patterns.length > 0) {
        try {
            child_process_1.execSync("node " + prettierPath() + " --single-quote --print-width 120 --list-different " + patterns.join(' '), {
                stdio: [0, 1, 2]
            });
        }
        catch (e) {
            process.exit(1);
        }
    }
}
function prettierPath() {
    var basePath = path.dirname(resolve.sync('prettier', { basedir: __dirname }));
    return path.join(basePath, 'bin-prettier.js');
}
