// @nrwl/schematics/src/collection/utility/lib-versions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.angularCliVersion = 'file:.angular_cli.tgz';
exports.angularVersion = '5.1.0';
exports.angularJsVersion = '1.6.6';
exports.ngrxVersion = '4.1.1';
exports.nxVersion = '0.6.19';
exports.schematicsVersion = '0.6.19';
exports.angularCliSchema = './node_modules/@nrwl/schematics/src/schema.json';
exports.latestMigration = '20180120-update-prettier';
exports.prettierVersion = '1.10.2';
exports.typescriptVersion = '2.5.3';
exports.rxjsVersion = '^5.5.2';
exports.devKitCoreVersion = '^0.0.28';
exports.libVersions = {
    angularVersion: exports.angularVersion,
    angularCliVersion: exports.angularCliVersion,
    angularJsVersion: exports.angularJsVersion,
    ngrxVersion: exports.ngrxVersion,
    nxVersion: exports.nxVersion,
    schematicsVersion: exports.schematicsVersion,
    prettierVersion: exports.prettierVersion,
    angularCliSchema: exports.angularCliSchema,
    latestMigration: exports.latestMigration,
    typescriptVersion: exports.typescriptVersion,
    rxjsVersion: exports.rxjsVersion,
    devKitCoreVersion: exports.devKitCoreVersion
};
